compiler/rustc_expand/src/mbe/metavar_expr.rs

```diff
@@ -0,0 +1,180 @@
+                    token.span,
+                    &format!(
+                        "expected meta-variable expression, found `{}`",
+                        pprust::token_to_string(&token),
+                    ),
+                ));
+            }
+            _ => return Err(sess.span_diagnostic.struct_err("expected meta-variable expression"))
+        }
+    }
+
+    crate fn ident(&self) -> Option<&Ident> {
+        match self {
+            MetaVarExpr::Count(ident, _) | MetaVarExpr::Ignore(ident) => Some(&ident),
+            MetaVarExpr::Index(..) | MetaVarExpr::Length(..) => None,
+        }
+    }
+}
+
+/// Tries to convert a literal to an arbitrary type
+fn convert_literal<T>(lit: Lit, sess: &ParseSess, span: Span) -> PResult<'_, T>
+where
+    T: FromStr,
+    <T as FromStr>::Err: Display,
+{
+    if lit.suffix.is_some() {
+        let msg = "literal suffixes are not supported in meta-variable expressions";
+        return Err(sess.span_diagnostic.struct_span_err(span, msg));
+    }
+    lit.symbol.as_str().parse::<T>().map_err(|e| {
+        sess.span_diagnostic.struct_span_err(
+            span,
+            &format!("failed to parse meta-variable expression argument: {}", e),
+        )
+    })
+}
+
+/// Parse a meta-variable `count` expression: `count(ident[, depth])`
+fn parse_count<'sess>(
+    iter: &mut tokenstream::Cursor,
+    sess: &'sess ParseSess,
+    span: Span,
+) -> PResult<'sess, MetaVarExpr> {
+    let ident = parse_ident(iter, sess, span)?;
+    let depth = if try_eat_comma(iter) { Some(parse_depth(iter, sess, span)?) } else { None };
+    Ok(MetaVarExpr::Count(ident, depth))
+}
+
+/// Parses the depth used by index(depth) and length(depth).
+fn parse_depth<'sess>(
+    iter: &mut tokenstream::Cursor,
+    sess: &'sess ParseSess,
+    span: Span,
+) -> PResult<'sess, usize> {
+    let Some(tt) = iter.next() else { return Ok(0) };
+    let tokenstream::TokenTree::Token(token::Token {
+        kind: token::TokenKind::Literal(lit),
+        span: literal_span,
+    }) = tt else {
+        return Err(sess.span_diagnostic.struct_span_err(
+            span,
+            "meta-expression depth must be a literal"
+        ));
+    };
+    convert_literal::<usize>(lit, sess, literal_span)
+}
+
+/// Parses an generic ident
+fn parse_ident<'sess>(
+    iter: &mut tokenstream::Cursor,
+    sess: &'sess ParseSess,
+    span: Span,
+) -> PResult<'sess, Ident> {
+    let err_fn =
+        || sess.span_diagnostic.struct_span_err(span, "could not find an expected `ident` element");
+    if let Some(tt) = iter.next() {
+        match tt {
+            tokenstream::TokenTree::Token(token) => {
+                if let Some((elem, false)) = token.ident() {
+                    return Ok(elem);
+                }
+                let mut err = err_fn();
+                err.span_suggestion(
+                    token.span,
+                    &format!("Try removing `{}`", pprust::token_to_string(&token)),
```

Comment:
```suggestion
                    &format!("try removing `{}`", pprust::token_to_string(&token)),
```

Reply:
Also, this suggestion doesn't make sense in a general case, I'd report it only if the current token is `$` and the next token is an identifier (if that was the original goal).